### DISCUSSION TITLE
External linear axis - track synergistic with ABB robot & real-time data streaming

### DISCUSSION THREAD

**Discussion by fz0707** [visose/Robots#282] [Q&A] [answered]
Hi Vince,

I'm working with an ABB 6640 and an external linear axis (track). I have a couple of questions:

1. How do I integrate the external linear axis with the robot? You mentioned before: "To do this, add the 'External' input to the CreateTarget component and supply a list of 3 numbers, one per axis. You must manually specify the position of each axis per target. The simulation and the robot controller will not do 'inverse kinematics' on all 9 axes, only on the 6 axes of the robot." Does this mean the external axis isn't considered in forward and inverse kinematics? For instance, in 3D printing, how can we create a linear trajectory that moves the robot and the track together smoothly, so the robot keeps printing without any stops while following the track to maximize the printing range?

2. Our code sometimes extends to hundreds of thousands of lines, but ABB's limit seems to be around 45,000 lines per file. Is there a way to stream data from Grasshopper to the robot? I've heard it requires multi-tasking and a PC interface, but I'm not sure about the details.

Looking forward to your help.

Thanks a lot,  
Charles

**Comment by visose** [comment, marked as answer]
Hi,
> How do I integrate the external linear axis with the robot?

You can check the `Penn` library for an example of an ABB with a track.

> Does this mean the external axis isn't considered in forward and inverse kinematics? 

When you create the targets, you have to explicitly define on each target the values of your external axes. It will do forward kinematics to position the base of the robot given those values, then use inverse kinematics on the robot joints only to position the TCP.

You can use some Grasshopper logic to define the position of the track on each target. The simplest example would be (asuming the track goes in the X axis), to extract the X coordinate of each target and use that as the track location, but the base will move on every target and you won't be using the reach of the arm in that direction. Moving the base requires more energy, acceleration and precision is lower, etc. You might prefer the base to move to a position and remain static for the consecutive targets.

> Our code sometimes extends to hundreds of thousands of lines, but ABB's limit seems to be around 45,000 lines per file. 

You can use the Multi-file indices input of the Create program component.

**Reply by fz0707** [reply]
Many thanks for your reply, Vince.

I am wondering whether the robot and the track move simultaneously or if the robot first moves to the set position before starting its operation. If we assume the track can move quite smoothly, I believe that the robot can only perform a long trajectory without stopping if both the track and robot joints move simultaneously. Please let me know if my understanding is correct.

**Reply by visose** [reply]
The robot and track can move at the same time. You can use the Simulation component to preview the trajectory.

**Reply by fz0707** [reply]
Thanks, all clear!
I have another concern. I am planning to purchase an ABB robot and control it by Robots plugin. There are two options of controller - OmniCore or ICR5. Could you suggest which works best with Robots?

**Reply by visose** [reply]
Both work with Robots. I will release an update soon that creates .modx files directly for Omnicore, but you can also import the older .mod files.
